Missing migration from v1 config in pagerduty notification

The Pagerduty notification code was migrated in October 2020 from https://github.com/graylog-labs/graylog-plugin-pagerduty, and its config version bumped from `v1` to `v2` in the process.

There doesn't seem to be a migration for `v1` configs, so if a very old config is still present, this can lead to all kinds of issues whenever the entity is loaded.
In this case, a share dialog tried to load the `v1` config from the database and failed.

I think we need to add a v1 to v2 migration as a bug fix if that is possible. If it's not possible, we should fail more gracefully.

See https://github.com/Graylog2/graylog-plugin-integrations/blob/08a16fed7039c5e4c1bbdc269fa36dc458d91f1f/src/main/java/org/graylog/integrations/pagerduty/PagerDutyNotificationConfig.java#L47

Example v1 notification db entity:

```
{
  "_id": {
    "$oid": "5d4806962c1e3e45b94f394c"
  },
  "title": "Pager Duty",
  "description": "Migrated legacy alarm callback",
  "config": {
    "type": "pagerduty-notification-v1",
    "routing_key": "b07334643c6e4c509537dbdcd78d2587",
    "custom_incident": true,
    "key_prefix": "Graylog/",
    "client_name": "Graylog",
    "client_url": "https://xxx.yyy.com"
  }
```

The error looks like: (but it's generic, because the fallback object cannot be used directly):
```
2022-06-03T12:50:13.136Z ERROR [AnyExceptionClassMapper] Unhandled exception in REST resource
java.lang.RuntimeException: IOException encountered while reading from a byte array input stream
```

<details>
```
2022-06-03T12:50:13.136Z ERROR [AnyExceptionClassMapper] Unhandled exception in REST resource
java.lang.RuntimeException: IOException encountered while reading from a byte array input stream
	at org.mongojack.internal.stream.JacksonDBDecoder.decode(JacksonDBDecoder.java:67) ~[graylog.jar:?]
	at com.mongodb.DBDecoderAdapter.decode(DBDecoderAdapter.java:49) ~[graylog.jar:?]
	at com.mongodb.DBDecoderAdapter.decode(DBDecoderAdapter.java:29) ~[graylog.jar:?]
	at com.mongodb.operation.CommandResultArrayCodec.decode(CommandResultArrayCodec.java:52) ~[graylog.jar:?]
	at com.mongodb.operation.CommandResultDocumentCodec.readValue(CommandResultDocumentCodec.java:60) ~[graylog.jar:?]
	at org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:84) ~[graylog.jar:?]
	at org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:41) ~[graylog.jar:?]
	at org.bson.internal.LazyCodec.decode(LazyCodec.java:48) ~[graylog.jar:?]
	at org.bson.codecs.BsonDocumentCodec.readValue(BsonDocumentCodec.java:101) ~[graylog.jar:?]
	at com.mongodb.operation.CommandResultDocumentCodec.readValue(CommandResultDocumentCodec.java:63) ~[graylog.jar:?]
	at org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:84) ~[graylog.jar:?]
	at org.bson.codecs.BsonDocumentCodec.decode(BsonDocumentCodec.java:41) ~[graylog.jar:?]
	at com.mongodb.internal.connection.ReplyMessage.<init>(ReplyMessage.java:51) ~[graylog.jar:?]
	at com.mongodb.internal.connection.InternalStreamConnection.getCommandResult(InternalStreamConnection.java:413) ~[graylog.jar:?]
	at com.mongodb.internal.connection.InternalStreamConnection.receiveCommandMessageResponse(InternalStreamConnection.java:309) ~[graylog.jar:?]
	at com.mongodb.internal.connection.InternalStreamConnection.sendAndReceive(InternalStreamConnection.java:259) ~[graylog.jar:?]
	at com.mongodb.internal.connection.UsageTrackingInternalConnection.sendAndReceive(UsageTrackingInternalConnection.java:99) ~[graylog.jar:?]
	at com.mongodb.internal.connection.DefaultConnectionPool$PooledConnection.sendAndReceive(DefaultConnectionPool.java:450) ~[graylog.jar:?]
	at com.mongodb.internal.connection.CommandProtocolImpl.execute(CommandProtocolImpl.java:72) ~[graylog.jar:?]
	at com.mongodb.internal.connection.DefaultServer$DefaultServerProtocolExecutor.execute(DefaultServer.java:226) ~[graylog.jar:?]
	at com.mongodb.internal.connection.DefaultServerConnection.executeProtocol(DefaultServerConnection.java:269) ~[graylog.jar:?]
	at com.mongodb.internal.connection.DefaultServerConnection.command(DefaultServerConnection.java:131) ~[graylog.jar:?]
	at com.mongodb.internal.connection.DefaultServerConnection.command(DefaultServerConnection.java:123) ~[graylog.jar:?]
	at com.mongodb.operation.CommandOperationHelper.executeCommand(CommandOperationHelper.java:343) ~[graylog.jar:?]
	at com.mongodb.operation.CommandOperationHelper.executeCommand(CommandOperationHelper.java:334) ~[graylog.jar:?]
	at com.mongodb.operation.CommandOperationHelper.executeCommandWithConnection(CommandOperationHelper.java:220) ~[graylog.jar:?]
	at com.mongodb.operation.FindOperation$1.call(FindOperation.java:731) ~[graylog.jar:?]
	at com.mongodb.operation.FindOperation$1.call(FindOperation.java:725) ~[graylog.jar:?]
	at com.mongodb.operation.OperationHelper.withReadConnectionSource(OperationHelper.java:463) ~[graylog.jar:?]
	at com.mongodb.operation.FindOperation.execute(FindOperation.java:725) ~[graylog.jar:?]
	at com.mongodb.operation.FindOperation.execute(FindOperation.java:89) ~[graylog.jar:?]
	at com.mongodb.client.internal.MongoClientDelegate$DelegateOperationExecutor.execute(MongoClientDelegate.java:196) ~[graylog.jar:?]
	at com.mongodb.client.internal.MongoClientDelegate$DelegateOperationExecutor.execute(MongoClientDelegate.java:177) ~[graylog.jar:?]
	at com.mongodb.DBCursor.initializeCursor(DBCursor.java:989) ~[graylog.jar:?]
	at com.mongodb.DBCursor.hasNext(DBCursor.java:172) ~[graylog.jar:?]
	at org.mongojack.DBCursor.hasNext(DBCursor.java:330) ~[graylog.jar:?]
	at java.util.Iterator.forEachRemaining(Iterator.java:115) ~[?:1.8.0_312]
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801) ~[?:1.8.0_312]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482) ~[?:1.8.0_312]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472) ~[?:1.8.0_312]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708) ~[?:1.8.0_312]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:1.8.0_312]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:566) ~[?:1.8.0_312]
	at org.graylog.events.contentpack.facade.NotificationFacade.listEntityExcerpts(NotificationFacade.java:129) ~[graylog.jar:?]
	at org.graylog2.contentpacks.ContentPackService.lambda$listAllEntityExcerpts$3(ContentPackService.java:313) ~[graylog.jar:?]
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:422) ~[graylog.jar:?]
	at org.graylog2.contentpacks.ContentPackService.listAllEntityExcerpts(ContentPackService.java:313) ~[graylog.jar:?]
	at org.graylog.security.entities.EntityDependencyResolver.resolve(EntityDependencyResolver.java:69) ~[graylog.jar:?]
	at org.graylog.security.shares.EntitySharesService.checkMissingPermissionsOnDependencies(EntitySharesService.java:298) ~[graylog.jar:?]
	at org.graylog.security.shares.EntitySharesService.prepareShare(EntitySharesService.java:108) ~[graylog.jar:?]
	at org.graylog.security.rest.EntitySharesResource.prepareShare(EntitySharesResource.java:123) ~[graylog.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_312]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_312]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_312]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_312]
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52) ~[graylog.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124) ~[graylog.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167) ~[graylog.jar:?]
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:219) ~[graylog.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79) ~[graylog.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:469) ~[graylog.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:391) ~[graylog.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:80) ~[graylog.jar:?]
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:255) [graylog.jar:?]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248) [graylog.jar:?]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244) [graylog.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292) [graylog.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274) [graylog.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244) [graylog.jar:?]
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265) [graylog.jar:?]
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234) [graylog.jar:?]
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:680) [graylog.jar:?]
	at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpContainer.service(GrizzlyHttpContainer.java:356) [graylog.jar:?]
	at org.glassfish.grizzly.http.server.HttpHandler$1.run(HttpHandler.java:200) [graylog.jar:?]
	at com.codahale.metrics.InstrumentedExecutorService$InstrumentedRunnable.run(InstrumentedExecutorService.java:180) [graylog.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_312]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_312]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_312]
Caused by: com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "type" (class org.graylog.events.notifications.EventNotificationConfig$FallbackNotificationConfig), not marked as ignorable (0 known properties: ])
 at [Source: de.undercouch.bson4jackson.io.LittleEndianInputStream@53c536ab; pos: 135] (through reference chain: org.graylog.events.notifications.AutoValue_NotificationDto$Builder["config"]->org.graylog.events.notifications.EventNotificationConfig$FallbackNotificationConfig["type"])
	at com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException.from(UnrecognizedPropertyException.java:61) ~[graylog.jar:?]
	at com.fasterxml.jackson.databind.DeserializationContext.handleUnknownProperty(DeserializationContext.java:823) ~[graylog.jar:?]
	at com.fasterxml.jackson.databind.deser.std.StdDeserializer.handleUnknownProperty(StdDeserializer.java:1153) ~[graylog.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownProperty(BeanDeserializerBase.java:1589) ~[graylog.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownVanilla(BeanDeserializerBase.java:1567) ~[graylog.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:294) ~[graylog.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:189) ~[graylog.jar:?]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:161) ~[graylog.jar:?]
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:130) ~[graylog.jar:?]
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:97) ~[graylog.jar:?]
	at com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserializeWithType(AbstractDeserializer.java:254) ~[graylog.jar:?]
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeSetAndReturn(MethodProperty.java:167) ~[graylog.jar:?]
	at com.fasterxml.jackson.databind.deser.BuilderBasedDeserializer.vanillaDeserialize(BuilderBasedDeserializer.java:269) ~[graylog.jar:?]
	at com.fasterxml.jackson.databind.deser.BuilderBasedDeserializer.deserialize(BuilderBasedDeserializer.java:193) ~[graylog.jar:?]
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:3985) ~[graylog.jar:?]
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2343) ~[graylog.jar:?]
	at org.mongojack.internal.stream.JacksonDBDecoder.decode(JacksonDBDecoder.java:80) ~[graylog.jar:?]
	at org.mongojack.internal.stream.JacksonDBDecoder.decode(JacksonDBDecoder.java:64) ~[graylog.jar:?]
	... 77 more
```
</details>

--- a/src/main/java/org/graylog/integrations/IntegrationsModule.java
+++ b/src/main/java/org/graylog/integrations/IntegrationsModule.java
@@ -35,6 +35,7 @@ import org.graylog.integrations.inputs.paloalto9.PaloAlto9xInput;
 import org.graylog.integrations.ipfix.codecs.IpfixCodec;
 import org.graylog.integrations.ipfix.inputs.IpfixUdpInput;
 import org.graylog.integrations.ipfix.transports.IpfixUdpTransport;
+import org.graylog.integrations.migrations.V20220622071600_MigratePagerDutyV1;
 import org.graylog.integrations.notifications.types.SlackEventNotification;
 import org.graylog.integrations.notifications.types.SlackEventNotificationConfig;
 import org.graylog.integrations.notifications.types.microsoftteams.TeamsEventNotification;
@@ -172,6 +173,9 @@ public class IntegrationsModule extends PluginModule {
         bind(IamClientBuilder.class).toProvider(IamClient::builder);
         bind(CloudWatchLogsClientBuilder.class).toProvider(CloudWatchLogsClient::builder);
         bind(KinesisClientBuilder.class).toProvider(KinesisClient::builder);
+
+        // PagerDuty notification type fix
+        addMigration(V20220622071600_MigratePagerDutyV1.class);
     }
 
     /**

--- a/src/main/java/org/graylog/integrations/migrations/V20220622071600_MigratePagerDutyV1.java
+++ b/src/main/java/org/graylog/integrations/migrations/V20220622071600_MigratePagerDutyV1.java
@@ -1,0 +1,95 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.integrations.migrations;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.google.auto.value.AutoValue;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.model.Filters;
+import com.mongodb.client.model.Updates;
+import com.mongodb.client.result.UpdateResult;
+import org.bson.Document;
+import org.bson.conversions.Bson;
+import org.graylog.autovalue.WithBeanGetter;
+import org.graylog.integrations.pagerduty.PagerDutyNotificationConfig;
+import org.graylog2.database.MongoConnection;
+import org.graylog2.migrations.Migration;
+import org.graylog2.plugin.cluster.ClusterConfigService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.time.ZonedDateTime;
+
+public class V20220622071600_MigratePagerDutyV1 extends Migration {
+    private static final Logger LOG = LoggerFactory.getLogger(V20220622071600_MigratePagerDutyV1.class);
+    private static final String COLLECTION_NAME = "event_notifications";
+    private static final String PAGER_DUTY_V1 = "pagerduty-notification-v1";
+    private static final String TYPE_FIELD = "config.type";
+
+    private final MongoConnection mongoConnection;
+    private final ClusterConfigService clusterConfigService;
+
+    @Inject
+    public V20220622071600_MigratePagerDutyV1(MongoConnection mongoConnection, ClusterConfigService clusterConfigService) {
+        this.mongoConnection = mongoConnection;
+        this.clusterConfigService = clusterConfigService;
+    }
+
+    @Override
+    public ZonedDateTime createdAt() {
+        return ZonedDateTime.parse("2022-06-22t07:16Z");
+    }
+
+    /**
+     * This migration modifies the config.type field for {@link PagerDutyNotificationConfig} that were created before
+     * it was migrated into Graylog Core. Previous configurations had config.type 'pagerduty-notification-v1' and the
+     * type was updated to 'pagerduty-notification-v2' in the migration. No other fields were added/modified/deleted
+     * in the migration, so only the config.type field is updated.
+     */
+    @Override
+    public void upgrade() {
+        V20220622071600_MigratePagerDutyV1.MigrationCompletion completion = clusterConfigService.get(V20220622071600_MigratePagerDutyV1.MigrationCompletion.class);
+        if (completion != null) {
+            LOG.debug("Migration was already completed");
+            return;
+        }
+
+        final MongoCollection<Document> collection = mongoConnection.getMongoDatabase().getCollection(COLLECTION_NAME);
+        LOG.info("Updating '{}' collection.", COLLECTION_NAME);
+
+        Bson v1Filter = Filters.eq(TYPE_FIELD, PAGER_DUTY_V1);
+        Bson v2Update = Updates.set(TYPE_FIELD, PagerDutyNotificationConfig.TYPE_NAME);
+
+        LOG.info("Updating {} from {} to {}", TYPE_FIELD, PAGER_DUTY_V1, PagerDutyNotificationConfig.TYPE_NAME);
+        final UpdateResult updateTypeResult = collection.updateOne(v1Filter, v2Update);
+        LOG.info("Update result: {}", updateTypeResult);
+
+        clusterConfigService.write(MigrationCompletion.create());
+    }
+
+    @JsonAutoDetect
+    @AutoValue
+    @WithBeanGetter
+    public abstract static class MigrationCompletion {
+        @JsonCreator
+        public static V20220622071600_MigratePagerDutyV1.MigrationCompletion create() {
+            return new AutoValue_V20220622071600_MigratePagerDutyV1_MigrationCompletion();
+        }
+    }
+}

--- a/src/main/java/org/graylog/integrations/migrations/V20220622071600_MigratePagerDutyV1.java
+++ b/src/main/java/org/graylog/integrations/migrations/V20220622071600_MigratePagerDutyV1.java
@@ -77,7 +77,7 @@ public class V20220622071600_MigratePagerDutyV1 extends Migration {
         Bson v2Update = Updates.set(TYPE_FIELD, PagerDutyNotificationConfig.TYPE_NAME);
 
         LOG.info("Updating {} from {} to {}", TYPE_FIELD, PAGER_DUTY_V1, PagerDutyNotificationConfig.TYPE_NAME);
-        final UpdateResult updateTypeResult = collection.updateOne(v1Filter, v2Update);
+        final UpdateResult updateTypeResult = collection.updateMany(v1Filter, v2Update);
         LOG.info("Update result: {}", updateTypeResult);
 
         clusterConfigService.write(MigrationCompletion.create());

